File: Stepper.py
from numpy.random import random_integers
import torch
from torch.autograd import Variable
import torch.nn.functional as F
import torch.utils.data as Data
from torch.utils.data.dataset import Dataset
import torch_optimizer as optim
import random
from torch.utils.tensorboard import SummaryWriter

import numpy as np
import time

start_time = time.time()

stepperInput = []
stepperNext = []
latent = []

with open("/home/pau1o-hs/Documents/Database/XData.txt") as f:
    for line in f:
        inner_list = [elt.strip() for elt in line.split(' ')]

        if inner_list == ['']:
            continue

        converted =  np.asarray(inner_list, dtype=np.float64, order='C')
        stepperInput.append(converted)

with open("/home/pau1o-hs/Documents/Database/LatentVariables.txt") as f:
    for line in f:
        inner_list = [elt.strip() for elt in line.split(' ')]
        
        if inner_list == ['']:
            continue
        
        converted =  np.asarray(inner_list, dtype=np.float64, order='C')
        latent.append(converted)

device = torch.device("cuda")

stepperInput = np.append(stepperInput, latent, axis=1)
stepperNext = np.roll(stepperInput, -1, axis=0)

x = torch.tensor(stepperInput, dtype=torch.float).to(device)        # x data (tensor), shape=(100, 1)
y = torch.tensor(stepperNext,  dtype=torch.float).to(device)

means = x.mean(dim=1, keepdim=True)
stds = x.std(dim=1, keepdim=True)
normalized_input = (x - means) / stds

means = y.mean(dim=1, keepdim=True)
stds = y.std(dim=1, keepdim=True)
normalized_output = (y - means) / stds

# torch can only train on Variable, so convert them to Variable
# x, y = Variable(x), Variable(y)

class MyDataset(Dataset):
    def __init__(self, data, window=20):
        self.data = data
        self.window = window

    def __getitem__(self, index):
        x = self.data[index:index+self.window]
        return x

    def __len__(self):
        return len(self.data) - self.window

train = Data.TensorDataset(normalized_input, normalized_output)

dataSet = MyDataset(train)

train_loader = Data.DataLoader(dataSet, shuffle=True, batch_size=32)

class Net(torch.nn.Module):
    def __init__(self, n_feature, n_hidden, n_hidden2, n_output):
        super(Net, self).__init__()
        self.n_hidden = n_hidden
        self.n_hidden2 = n_hidden2
        self.n_output = n_output

        self.hidden  = torch.nn.LSTM(n_feature, n_hidden, 1, batch_first=True)   # hidden layer
        self.hidden2 = torch.nn.LSTM(n_hidden, n_hidden2, 1, batch_first=True)   # hidden layer
        self.predict = torch.nn.Linear(n_hidden2, n_output)   # output layer

        self.gru = torch.nn.GRU(n_hidden, n_hidden)

    def forward(self, x, future=10):
        outputs = []
        n_batch = x.size(0)
        n_samples = x.size(1)

        h_t = torch.zeros(1, n_batch, self.n_hidden, dtype=torch.float32, device=device)
        c_t = torch.zeros(1, n_batch, self.n_hidden, dtype=torch.float32, device=device)
        h_t2 = torch.zeros(1, n_batch, self.n_hidden, dtype=torch.float32, device=device)
        c_t2 = torch.zeros(1, n_batch, self.n_hidden, dtype=torch.float32, device=device)
        
        h_t, c_t = self.hidden(x, (h_t, c_t))
        h_t = F.relu(h_t)

        h_t2, c_t2 = self.hidden2(h_t, (h_t2, c_t2))
        h_t2 = F.relu(h_t2)

        output = self.predict(h_t2)
        outputs.append(output)

        # output = torch.cat(output, dim=0).split(1, dim=1).split(1, dim=2)
        # print(output.shape)

        # for i in range(future):
        #     print(output.split(1, dim=0)[-1].shape)
        #     h_t, c_t = F.relu(self.hidden(output.split(1, dim=0)[-1], (h_t, c_t)))
        #     h_t2, c_t2 = F.relu(self.hidden(h_t, (h_t2, c_t2)))
        #     output = self.predict(h_t2)
        #     outputs.append(output)

        outputs = torch.cat(outputs, dim=1)
        return output

        # x = F.relu(self.hidden(x))      # activation function for hidden layer
        # x = F.relu(self.hidden2(x))     # activation function for hidden layer
        # x = self.predict(x)             # linear output
        # return x

net = Net(n_feature=24+32, n_hidden=512, n_hidden2=512, n_output=24+32).to(device)     # define the network

# optimizer = torch.optim.SGD(net.parameters(), lr=0.1)
optimizer = optim.RAdam(net.parameters(), lr=0.001)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1000, gamma=0.99)
loss_func = torch.nn.MSELoss()  # this is for regression mean squared loss

writer = SummaryWriter()

# train the network
for t in range(1):
    epochLoss = 0.0

    for batch_idx, (data, target) in enumerate(train_loader):    
        
        data.to(device), target.to(device) 

        prediction = net(data)     # input x and predict based on x
        # print(prediction.shape)

        delta = target - data
        loss = loss_func(prediction, delta)     # must be (1. nn output, 2. target)

        optimizer.zero_grad()   # clear gradients for next train
        loss.backward()         # backpropagation, compute gradients
        optimizer.step()        # apply gradients
        scheduler.step()

        epochLoss += loss * prediction.size(0)
        # print("batchComplete")

    if t % 500 == 0:
        print(t, epochLoss.item())
    
    writer.add_scalar('Python Stepper Loss', epochLoss, t)

torch.set_printoptions(precision=6)
np.set_printoptions(precision=7, floatmode='fixed', suppress=True)

# ar_hidden = net.initHidden(20)
# prediction, ar_hidden = net(normalized_input[0], ar_hidden)

# print(x[0].detach().numpy())
# print()
# print(y[0].detach().numpy())
# print()
# print(prediction.detach().numpy())

with torch.no_grad():
    inputx = torch.randn(1, 20, 56).to(device)

    # Export the model
    torch.onnx.export(net,                              # model being run
                    inputx,                  # model dummy input (or a tuple for multiple inputs)
                    "stepper.onnx",                     # where to save the model (can be a file or file-like object)
                    export_params=True,                 # store the trained parameter weights inside the model file
                    opset_version=9,                    # the ONNX version to export the model to
                    do_constant_folding=True,           # whether to execute constant folding for optimization
                    input_names = ['x', 'h0', 'c0'],    # the model's input names
                    output_names = ['y'],               # the model's output names
                    dynamic_axes={'x': {0: 'sequence'}, 'y': {0: 'sequence'}}
                  )

# with open('/home/pau1o-hs/Documents/NNWeights/Stepper.txt', "w+") as f:
#     np.savetxt(f, net.hidden.weight.cpu().detach().numpy().transpose(), delimiter="\n")
#     np.savetxt(f, net.hidden.bias.cpu().detach().numpy(), delimiter="\n")

#     np.savetxt(f, net.hidden2.weight.cpu().detach().numpy().transpose(), delimiter="\n")
#     np.savetxt(f, net.hidden2.bias.cpu().detach().numpy(), delimiter="\n")

#     np.savetxt(f, net.predict.weight.cpu().detach().numpy().transpose(), delimiter="\n")
#     np.savetxt(f, net.predict.bias.cpu().detach().numpy(), delimiter="\n")

    # np.savetxt(f, np.append(net.gru.all_weights[0][0].cpu().detach().numpy(), net.gru.all_weights[0][2].view(1536, 1).cpu().detach().numpy(), axis=1).transpose(), delimiter="\n")    
    # np.savetxt(f, np.append(net.gru.all_weights[0][1].cpu().detach().numpy(), net.gru.all_weights[0][3].view(1536, 1).cpu().detach().numpy(), axis=1).transpose(), delimiter="\n")

# nWeights = torch.numel(net.hidden.weight)
# nWeights += torch.numel(net.hidden.bias)

# nWeights += torch.numel(net.hidden2.weight)
# nWeights += torch.numel(net.hidden2.bias)

# nWeights += torch.numel(net.predict.weight)
# nWeights += torch.numel(net.predict.bias)

# nWeights += torch.numel(net.gru.all_weights[0][0])
# nWeights += torch.numel(net.gru.all_weights[0][1])
# nWeights += torch.numel(net.gru.all_weights[0][2])
# nWeights += torch.numel(net.gru.all_weights[0][3])

# print(net.hidden.weight.shape)
# print(net.hidden.bias.shape)
# print(net.hidden2.weight.shape)
# print(net.hidden2.bias.shape)
# print(net.predict.weight.shape)
# print(net.predict.bias.shape)
# print(nWeights)
# print(net.gru.all_weights[0][0][512])
# print(net.gru.all_weights[0][0].size(0), net.gru.all_weights[0][0].size(1))
# print(net.gru.all_weights[0][1].size(0), net.gru.all_weights[0][1].size(1))
# print(net.gru.all_weights[0][2].size(0), net.gru.all_weights[0][2].size(0))
# print(net.gru.all_weights[0][3].size(0), net.gru.all_weights[0][3].size(0))

print("Runtime: %s minutes" % ((time.time() - start_time) / 60))

File: Stepper.cpp
#include "CustomLib.hpp"

// C++ neural network model defined with value semantics
struct Net : torch::nn::Module
{
    Net(torch::Device device, int64_t n_feature, int64_t n_hidden1, int64_t  n_hidden2, int64_t n_output)
    {
        this->fc1     = register_module("fc1",     torch::nn::Linear(n_feature, n_hidden1));
        this->fc2     = register_module("fc2",     torch::nn::Linear(n_hidden1, n_hidden2));
        this->predict = register_module("predict", torch::nn::Linear(n_hidden2, n_output));

        this->fc1->to(device);
        this->fc2->to(device);
        this->predict->to(device);
    }

    torch::Tensor forward(torch::Tensor feature)
    {
        feature = torch::relu(this->fc1(feature));
        feature = torch::relu(this->fc2(feature));
        feature = this->predict(feature);
        return feature;
    }

    torch::nn::Linear fc1 = nullptr;
    torch::nn::Linear fc2 = nullptr;
    torch::nn::Linear predict = nullptr;
};

int main()
{
    auto start = chrono::high_resolution_clock::now();

	cout << "[Device: " << at::cuda::getDeviceProperties(0)->name << "]" << endl;
    cout << "[Count: " << torch::cuda::device_count() << "]" << endl << endl;

    // VARIABLES
    const int64_t n_feature  = 56;
    const int64_t n_hidden1  = 512;
    const int64_t n_hidden2  = 512;
    const int64_t n_output   = 56;
    const int64_t batch_size = 32;

    // INIT NN
    auto model = Net(device, n_feature, n_hidden1, n_hidden2, n_output);
    model.to(device);

    // DATASET
    vector<torch::Tensor> x, xNorm, yNorm, xFinal, yFinal;
    LoadData("Features", x, true);

	// NORMALIZE DATA
    for (int i = 0; i < (int) x.size(); i++)
    {
        auto mean = x[i].mean(1, true);
        auto std  = x[i].std(1, true).reshape({x[i].size(0), 1});

        xNorm.push_back((x[i] - mean) / std);
        yNorm.push_back(xNorm[i].roll(-1, 0).clone());
    }

    // cout << xNorm[0].size(0) << endl;

    // // POP LAST ROW
    // for (int i = 0; i < (int) xNorm.size(); i++)
    // {
    //     auto options = torch::TensorOptions().dtype(torch::kFloat);
    //     int n = (int) xNorm[i].size(0) - 1;
    //     int m = (int) xNorm[i].size(1);

    //     xFinal.push_back(torch::zeros({ n, m }, options).to(device));
    //     yFinal.push_back(torch::zeros({ n, m }, options).to(device));

	// 	for (int j = 0; j < n; j++)
    //     {
	// 		xFinal[i].slice(0, j, j + 1) = xNorm[i][j].clone();
	// 		yFinal[i].slice(0, j, j + 1) = yNorm[i][j].clone();
    //     }
    // }
    
    // cout << xFinal[0].size(0) << endl;

    auto dataset = CustomDataset(xNorm, yNorm, 20, 0).map(torch::data::transforms::Stack<>());

    // DATA LOADER
    auto data_loader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>
    (std::move(dataset), torch::data::DataLoaderOptions().batch_size(batch_size).workers(4));

    // OPTIMIZER
    torch::optim::AdamW optimizer(model.parameters(), 0.001);
	torch::optim::StepLR scheduler(optimizer, 1000, 0.99);

    cout << "Training Stepper...\n";
	
    for (size_t epoch = 1; epoch <= 10000; epoch++)
    {
        int batchId = 0;
        float epochLoss = 0;

        // Iterate the data loader to yield batches from the dataset.
        for (auto &batch : *data_loader)
        {
            batch.data = batch.data.to(device);
            batch.target = batch.target.to(device);
            
            // cout << batch.data.size(0) << ' ' << batch.data.size(1) << ' ' << batch.data.size(2) << endl;

            // Execute the model on the input data.
            torch::Tensor prediction = model.forward(batch.data);
            
            // Compute a loss value to judge the prediction of our model.
            torch::Tensor loss = torch::nn::MSELoss()(prediction, batch.target);

            // Reset gradients.
            optimizer.zero_grad();

            // Compute gradients of the loss w.r.t. the parameters of our model.
            loss.backward();

            // Update the parameters based on the calculated gradients.
            optimizer.step();
            scheduler.step();

            epochLoss += prediction.size(0) * loss.item<float>();
            
            // // Output the loss and checkpoint every 100 batches.
            // if (++batchId % 9 == 0 && epoch % 1000 == 0)
            // {
            //     // std::cout << "Epoch: " << epoch << " | Loss: " << loss.item<float>() << std::endl;
            //     // Serialize your model periodically as a checkpoint.
            //     // torch::save(model, "Decompressor.pt");
            // }
        }

        if (epoch % 1000 == 0)
        {
            cout << "Epoch: " << epoch << " | Loss: " << epochLoss << endl;
        }
    }

    std::cout << "Finished\n\n";

    // torch::Tensor prediction = model.forward(x[0]);
    // cout << prediction << endl;
    // cout << prediction[0].transpose(0, 0) << endl;

	// SAVE WEIGHTS
	ofstream weights("/home/pau1o-hs/Documents/NNWeights/Stepper.txt");
	torch::Tensor weightData;

	// FC1
	weightData = model.fc1->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc1->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// FC2
	weightData = model.fc2->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.fc2->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	// PREDICT
	weightData = model.predict->weight.cpu().detach().transpose(0, 1).flatten();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

	weightData = model.predict->bias.cpu().detach();
	for (int i = 0; i < (int) weightData.size(0); i++) weights << weightData[i].item<float>() << endl;

    // RUNTIME
    auto stop = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::seconds>(stop - start);
    cout << "Runtime: " << (float) (duration.count() / 60.0f) << " minutes" << endl;
}